packages/pds/src/sequencer/outbox.ts

```diff
@@ -0,0 +1,111 @@
+import { AsyncBuffer, AsyncBufferFullError } from '@atproto/common'
+import Sequencer, { RepoAppendEvent } from '.'
+
+export type OutboxOpts = {
+  maxBufferSize: number
+}
+
+export class Outbox {
+  caughtUp = false
+  lastSeen = -1
+
+  cutoverBuffer: RepoAppendEvent[]
+  outBuffer: AsyncBuffer<RepoAppendEvent>
+
+  constructor(public sequencer: Sequencer, opts: Partial<OutboxOpts> = {}) {
+    const { maxBufferSize = 500 } = opts
+    this.cutoverBuffer = []
+    this.outBuffer = new AsyncBuffer<RepoAppendEvent>(maxBufferSize)
+  }
+
+  // event stream occurs in 3 phases
+  // 1. backfill events: events that have been added to the DB since the last time a connection was open.
+  // The outbox is not yet listening for new events from the sequencer
+  // 2. cutover: the outbox has caught up with where the sequencer purports to be,
+  // but the sequencer might already be halfway through sending out a round of updates.
+  // Therefore, we start accepting the sequencer's events in a buffer, while making our own request to the
+  // database to ensure we're caught up. We then dedupe the query & the buffer & stream the events in order
+  // 3. streaming: we're all caught up on historic state, so the sequencer outputs events and we
+  // immediately yield them
+  async *events(backfillFrom?: string): AsyncGenerator<RepoAppendEvent> {
+    // catch up as much as we can
+    if (backfillFrom) {
+      for await (const evt of this.getBackfill(backfillFrom)) {
+        yield evt
+        this.lastSeen = evt.seq
+      }
+    } else {
+      // if not backfill, we don't need to cutover, just start streaming
+      this.caughtUp = true
+    }
+
+    // streams updates from sequencer, but buffers them for cutover as it makes a last request
+    this.sequencer.on('events', (evts) => {
+      if (this.caughtUp) {
+        this.outBuffer.pushMany(evts)
+      } else {
+        this.cutoverBuffer = [...this.cutoverBuffer, ...evts]
+      }
+    })
+
+    const cutover = async () => {
+      // only need to perform cutover if we've been backfilling
+      if (backfillFrom) {
+        const cutoverEvts = await this.sequencer.requestSeqRange({
+          earliestSeq: this.lastSeen,
+        })
+        this.outBuffer.pushMany(cutoverEvts)
+        // dont worry about dupes, we ensure order on yield
+        this.outBuffer.pushMany(this.cutoverBuffer)
```

Comment:
Sort of working this out to myself.  On yield forward movement is always guaranteed by the `evt.seq > this.lastSeen` check.  So the thing to look out for would not be dupes, but instead skipping over some events. I.e. `1, 5, 4, 5, 6` would be a bad sequence of items in `cutoverEvents + this.cutoverBuffer`, since it would cause event `4` to get skipped over.

And that wont happen because... `this.cutoverBuffer` is inclusive of `cutoverEvts` and they are both in order?  Does that sound accurate?

Reply:
Yup exactly right on the first part.

both `cutoverBuffer` & `cutoverEvts` are in order. Not necessarily inclusive, but `cutoverBuffer` is strictly >= `cutoverEvt`

so for instance:
cutoverEvts: 1,2,3,4,5
cutoverBuffer: 3,4,5,6,7

pushed to out buffer as: 1,2,3,4,5,3,4,5,6,7
yielded to consumer as: 1,2,3,4,5,6,7

Reply:
Sweet, thanks!